### PR TITLE
Auto downloading JDK when none is present

### LINKED ISSUE #91: Automatically download java if not available on the system
Other extensions in VS code also try to do this, avoid asking the use to setup external stuff.

I see 3 cases:

1. User has a `JAVA_HOME` defined that points to a java 11+ runtime => we do nothing.
2. User has a `JAVA_HOME` defined that points to a older version of java => we download a jdk and use that.
3. User has *no* `JAVA_HOME` defined => we download a jdk and use that.

I propose we automatically download a LTS release from [Eclipse Adoptium](https://adoptium.net/) (the followup of the adopt-openjdk effort).

### PR DESCRIPTION
**Pull request by DavyLandman**
Fixes #91 

afbeelding
if user clicks automatic download:
afbeelding
after accepting license:
afbeelding
(the list of selections is based on OS and cpu architecture)
then a progressbar appears during the (200MB) download
afbeelding
when all is done, a last message is printed:
afbeelding

### CODE REVIEW

**Comment by DavyLandman** [comment]
Can you explain how this influences the java 11 branch?

I've been running rascal with a java 11 jdk for a long time.

**Comment by jurgenvinju** [comment]
I've never successfully run a Java 8 version of rascal on a Java 11 jdk; so I've no data about this and this I would be hesitant. I've done it years ago with crashes during unit testing as the biggest problem. Issues were related to different handling of System.err. Maybe that has been resolved. However I would say at least run our test suite with that setup before throwing it onto possibly newbie users. Developing DSLs only covers so much of the Rascal and vallang code...

**Comment by DavyLandman** [comment]
just ran it, the only test that fails is the m3 one (that is doing stuff with jar readers)

```
[ERROR] Tests run: 5863, Failures: 0, Errors: 3, Skipped: 203, Time elapsed: 167.012 s <<< FAILURE! - in org.rascalmpl.test.AllSuiteParallel
[ERROR] junitM3RemainedTheSame: <2656,197>(org.rascalmpl.test.infrastructure.RascalJUnitParallelRecursiveTestRunner$ModuleTester)  Time elapsed: 0.008 s  <<< ERROR!
org.rascalmpl.exceptions.Throw: |std:///lang/rascal/tests/library/lang/java/m3/BasicM3Tests.rsc|:167,13: "typeDependency has different value"
        at org.rascalmpl.test.infrastructure.RascalJUnitParallelRecursiveTestRunner$ModuleTester.runTests(RascalJUnitParallelRecursiveTestRunner.java:225)
        at org.rascalmpl.test.infrastructure.RascalJUnitParallelRecursiveTestRunner$ModuleTester.run(RascalJUnitParallelRecursiveTestRunner.java:206)

[ERROR] hamcrestJarM3RemainedTheSame: <4269,169>(org.rascalmpl.test.infrastructure.RascalJUnitParallelRecursiveTestRunner$ModuleTester)  Time elapsed: 0 s  <<< ERROR!
org.rascalmpl.exceptions.Throw: |std:///lang/java/m3/Core.rsc|:102,0: Java("IllegalArgumentException","")
        at org.rascalmpl.test.infrastructure.RascalJUnitParallelRecursiveTestRunner$ModuleTester.runTests(RascalJUnitParallelRecursiveTestRunner.java:225)
        at org.rascalmpl.test.infrastructure.RascalJUnitParallelRecursiveTestRunner$ModuleTester.run(RascalJUnitParallelRecursiveTestRunner.java:206)

[ERROR] junitASTsRemainedTheSame: <3352,203>(org.rascalmpl.test.infrastructure.RascalJUnitParallelRecursiveTestRunner$ModuleTester)  Time elapsed: 0 s  <<< ERROR!
java.lang.Exception:
Test lang::rascal::tests::library::lang::java::m3::BasicM3Tests::junitASTsRemainedTheSame failed due to
        test returned false
```

**Comment by DavyLandman** [comment]
Java 17 is causing some more subtle crashes:

```
[warn] |lib://typepal/analysis/typepal/ISolver.rsc|(35,49,<3,0>,<3,49>): Could not load analysis::typepal::ConfigurableScopeGraph due to: trying to construct a type from an unsupported type symbol: sort("ANONYMOUS_OCCURRENCE"), with this r
epresentation: Symbol = sort(str name) at |lib://typepal/analysis/typepal/ISolver.rsc|(35,49,<3,0>,<3,49>)
java.lang.IllegalArgumentException: trying to construct a type from an unsupported type symbol: sort("ANONYMOUS_OCCURRENCE"), with this representation: Symbol = sort(str name)
        at io.usethesource.vallang.type.TypeFactory$TypeValues.fromSymbol(TypeFactory.java:1098)
        at io.usethesource.vallang.type.TypeFactory.fromSymbol(TypeFactory.java:729)
        at org.rascalmpl.types.TypeReifier.symbolToType(TypeReifier.java:178)
        at org.rascalmpl.values.RascalValueFactory.reifiedType(RascalValueFactory.java:389)
        at org.rascalmpl.semantics.dynamic.Expression$AsType.interpret(Expression.java:1087)
        at org.rascalmpl.semantics.dynamic.Expression$Bracket.interpret(Expression.java:330)
        at org.rascalmpl.semantics.dynamic.Expression$GetAnnotation.interpret(Expression.java:965)
        at org.rascalmpl.semantics.dynamic.Declaration$Variable.interpret(Declaration.java:136)
        at org.rascalmpl.semantics.dynamic.Toplevel$GivenVisibility.interpret(Toplevel.java:35)
        at org.rascalmpl.semantics.dynamic.Module$Default.interpret(Module.java:79)
        at org.rascalmpl.semantics.dynamic.Import.loadModule(Import.java:319)
        at org.rascalmpl.semantics.dynamic.Import.extendCurrentModule(Import.java:270)
        at org.rascalmpl.semantics.dynamic.Import$Extend.interpret(Import.java:193)
        at org.rascalmpl.semantics.dynamic.Import.evalImport(Import.java:477)
        at org.rascalmpl.semantics.dynamic.Import.parseModuleAndFragments(Import.java:447)
        at org.rascalmpl.interpreter.Evaluator.parseModuleAndFragments(Evaluator.java:1435)
        at org.rascalmpl.interpreter.Evaluator.parseModuleAndFragments(Evaluator.java:1429)
        at org.rascalmpl.semantics.dynamic.Import.buildModule(Import.java:378)
        at org.rascalmpl.semantics.dynamic.Import.loadModule(Import.java:306)
        at org.rascalmpl.semantics.dynamic.Import.extendCurrentModule(Import.java:270)
        at org.rascalmpl.semantics.dynamic.Import$Extend.interpret(Import.java:193)
```

**Comment by PaulKlint** [comment]
@DavyLandman: this perceived Java 17 error is most likely caused by an older version of TypePal.

**Comment by jurgenvinju** [comment]
The M3 differences are expected; different versions of `rt.jar` may lead to different method resolutions, for example. Possibly also type resolutions? But have to look into the exact diff to see what is going on.

**Comment by jurgenvinju** [comment]
@PaulKlint this typepal error also happens when vallang can not load the type symbol implementations properly because the classpath is broken. `sort("ANONYMOUS_OCCURRENCE") ` is simply the very first `Symbol` that is  loaded if we load the type-checker.

@DavyLandman the `types.config` file is loaded using reflection in a way that seems incompatible with Java 17's module semantics? (an hypothesis)

**Comment by DavyLandman** [comment]
I'm now working on java 8 support, so for now we require java 8, and later we can switch to java 11.

**Comment by DavyLandman** [comment]
And yes, the reflection access in java 17 is much more restricted. we need to become actual java modules to deal with that.

**Comment by jurgenvinju** [comment]
Can we also do this when the current JDK is not of the right version?

File: rascal-vscode-extension/src/auto-jvm/downloaders.ts [hunk 1]
```diff
@@ -211,7 +215,7 @@ export async function fetchAndUnpackCorretto(arch: CorrettoArchitectures, platfo
 
 export async function fetchAndUnpackTemurin(arch: TemurinArchitectures, platform: TemurinPlatforms, jdk11Release: string, mainJVMPath: string, progress: ProgressFunc): Promise<string> {
     // https://api.adoptium.net/v3/binary/version/jdk-11.0.13%2B8/linux/x64/jdk/hotspot/normal/eclipse?project=jdk
-    const url = `https://api.adoptium.net/v3/binary/version/jdk-${encodeURIComponent(jdk11Release)}/${platform}/${arch}/jdk/hotspot/normal/eclipse?project=jdk`;
+    const url = `https://api.adoptium.net/v3/binary/version/${encodeURIComponent(jdk11Release)}/${platform}/${arch}/jdk/hotspot/normal/eclipse?project=jdk`;
```
Comment (jurgenvinju) [inline rascal-vscode-extension/src/auto-jvm/downloaders.ts:218]: perhaps rename the variable? jdk11Release to jdkRelease?

**Comment by DavyLandman** [comment]
> Can we also do this when the current JDK is not of the right version?

That is exactly what is happening now. It detects all available jvms, and checks their version. If not the right one, we'll show this dialog.

File: rascal-vscode-extension/src/auto-jvm/downloaders.ts (hunk 1 above)
Comment (DavyLandman) [reply rascal-vscode-extension/src/auto-jvm/downloaders.ts:218]: I though I got them all. I'll take care of this one as well.

**Comment by jurgenvinju** [comment]
Super. then let's merge this!

**Comment by DavyLandman** [comment]
Thanks, I have 1 thing to fix.

**Pull request merged**